src/markdown/generator.ml

```diff
@@ -0,0 +1,310 @@
+open Odoc_document
+open Types
+open Doctree
+
+module Markup = struct
+  type t =
+    | Inline of string list
+    | Block of t list
+    | Concat of t list
+    | Break
+    | Anchor of string
+    | String of string
+    | Env of indent * string * t * string
+    | Command of string * t
+    | OpenSqBracket
+    | CloseSqBracket
+    | OpenParenthesis
+    | CloseParenthesis
+
+  and indent = Any
+
+  let env indent s s' t = Env (indent, s, t, s')
+
+  let noop = Concat []
+
+  let break = Break
+
+  let open_sq_bracket, close_sq_bracket = (OpenSqBracket, CloseSqBracket)
+
+  let open_parenthesis, close_parenthesis = (OpenParenthesis, CloseParenthesis)
+
+  let append t1 t2 =
+    match (t1, t2) with
+    | Concat l1, Concat l2 -> Concat (l1 @ l2)
+    | Concat l1, e2 -> Concat (l1 @ [ e2 ])
+    | e1, Concat l2 -> Concat (e1 :: l2)
+    | e1, e2 -> Concat [ e1; e2 ]
+
+  let ( ++ ) = append
+
+  let concat = List.fold_left ( ++ ) (Concat [])
+
+  let rec intersperse ~sep = function
+    | [] -> []
+    | [ h ] -> [ h ]
+    | h1 :: (_ :: _ as t) -> h1 :: sep :: intersperse ~sep t
+
+  let list ?(sep = Concat []) l = concat @@ intersperse ~sep l
+
+  let str fmt = Format.ksprintf (fun s -> String s) fmt
+
+  let escaped fmt = Format.ksprintf (fun s -> String s) fmt
+
+  let command s t = Command (s, t)
+
+  let rec pp fmt t =
+    match t with
+    | Inline s -> Format.fprintf fmt "%s" (String.concat "" s)
+    | Block b ->
+        let inner = function
+          | [] -> ()
+          | [ x ] -> Format.fprintf fmt "%a" pp x
+          | x :: xs -> Format.fprintf fmt "%a@\n%a" pp x pp (Block xs)
+        in
+        inner b
+    | Concat l -> pp_many fmt l
+    | Break -> Format.fprintf fmt "@\n"
+    | Anchor s -> Format.fprintf fmt "<a id=\"%s\"></a>" s
+    | String s -> Format.fprintf fmt "%s" s
+    | Env (indent, s, t, s') ->
+        Format.fprintf fmt "@[%s%s%a@]@,%s@]"
+          (match indent with Any -> "")
+          s pp t s'
+    | Command (s, t) -> Format.fprintf fmt "@[%s%a@]" s pp t
+    | OpenSqBracket -> Format.fprintf fmt "["
+    | CloseSqBracket -> Format.fprintf fmt "]"
+    | OpenParenthesis -> Format.fprintf fmt "("
+    | CloseParenthesis -> Format.fprintf fmt ")"
+
+  and pp_many fmt l = List.iter (pp fmt) l
+end
+
+open Markup
+
+let entity e =
+  match e with "#45" -> escaped "-" | "gt" -> str ">" | s -> str "&%s;" s
+
+let raw_markup (_ : Raw_markup.t) = noop
+
+let style (style : style) content =
+  match style with
+  | `Bold -> command "**" (content ++ str "**")
+  | `Italic | `Emphasis -> command "_" (content ++ str "_")
+  | `Superscript -> command "<sup>" content
+  | `Subscript -> command "<sub>" content
```

Comment:
>Aren't the super/sub cases missing a closing tag here ?

The existing one should work well.

>Do you intend to try to do something about emphasis inside emphasis ?

No, because it would zero down to an emphasised statement. ( _Forexample _this_._)

Reply:
This is not fixed. The closing tag is missing: (`</sup>`)
```
<sup>test
```